File: app/src/main/java/com/amandariu/tagger/demo/main/MainActivity.java
package com.amandariu.tagger.demo.main;

import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.support.test.espresso.IdlingResource;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.amandariu.tagger.ITag;
import com.amandariu.tagger.TaggerActivity;
import com.amandariu.tagger.demo.R;
import com.amandariu.tagger.demo.Injection;
import com.amandariu.tagger.demo.utils.AlertUtils;
import com.amandariu.tagger.demo.utils.AndroidUtils;
import com.amandariu.tagger.demo.utils.EspressoIdlingResource;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements MainContract.View, View.OnClickListener {

    private static final String TAG = MainActivity.class.getSimpleName();

    private MainContract.Presenter mPresenter;
    //
    // View components
    private SimpleTagListAdapter mAvailableTagsAdapter;
    private SimpleTagListAdapter mSelectedTagsAdapter;
    private ViewGroup mViewLoading = null;
    private TextView mNoAvailableTags;
    private TextView mNoSelectedTags;
    //
    // State
    private List<? extends ITag> mAvailableTags = new ArrayList<>();
    private List<ITag> mSelectedTags = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //
        // Create the presenter
        new MainPresenter(Injection.provideTagsRepository(getApplicationContext()), this);
        //
        // Initialize View Components
        //
        // Get Tags from API
        Button btnGetApi = findViewById(R.id.btn_api);
        btnGetApi.setOnClickListener(this);
        //
        // Get Tags from Database
        Button btnGetDb = findViewById(R.id.btn_database);
        btnGetDb.setOnClickListener(this);
        //
        // Open the Tag Selector
        Button btnSelectTags = findViewById(R.id.btn_selectTags);
        btnSelectTags.setOnClickListener(this);
        //
        // Display Available Tags
        mNoAvailableTags = findViewById(R.id.txt_noAvailableTags);
        RecyclerView availList = findViewById(R.id.list_availableTags);
        mAvailableTagsAdapter = new SimpleTagListAdapter(new ArrayList<ITag>());
        availList.setAdapter(mAvailableTagsAdapter);
        Button clearAvailableTags = findViewById(R.id.btn_clearAvailableTags);
        clearAvailableTags.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mAvailableTags.clear();
                setAvailableTags(mAvailableTags);
            }
        });
        //
        // Display Selected Tags
        mNoSelectedTags = findViewById(R.id.txt_noSelectedTags);
        RecyclerView selectList = findViewById(R.id.list_selectedTags);
        mSelectedTagsAdapter = new SimpleTagListAdapter(new ArrayList<ITag>());
        selectList.setAdapter(mSelectedTagsAdapter);
        Button clearSelectedTags = findViewById(R.id.btn_clearSelectedTags);
        clearSelectedTags.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSelectedTags.clear();
                setSelectedTags(mSelectedTags);
            }
        });
        //
        // Loading indicator
        mViewLoading = findViewById(R.id.view_loading);

        if (savedInstanceState != null) {
            //
            // Restore any selected tags saved during configuration change.
            Parcelable[] pSelTags = savedInstanceState
                    .getParcelableArray(TaggerActivity.ARG_SELECTED_TAGS);
            if (pSelTags != null && pSelTags.length > 0) {
                mSelectedTags.clear();
                for (Parcelable t : pSelTags) {
                    mSelectedTags.add((ITag)t);
                }
                setSelectedTags(mSelectedTags);
            }
            //
            // Restore available tags saved during configuration change
            Parcelable[] pAvaTags = savedInstanceState
                    .getParcelableArray(TaggerActivity.ARG_AVAILABLE_TAGS);
            if (pAvaTags != null && pAvaTags.length > 0) {
                List<ITag> tags = new ArrayList<>(pAvaTags.length);
                for (Parcelable t : pAvaTags) {
                    tags.add((ITag) t);
                }
                setAvailableTags(tags);
            }
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        //
        // Save selected tags
        if (mSelectedTags.size() > 0) {
            outState.putParcelableArray(
                    TaggerActivity.ARG_SELECTED_TAGS,
                    mSelectedTags.toArray(new ITag[mSelectedTags.size()]));
        }
        //
        // Save available tags
        if (mAvailableTags != null && mAvailableTags.size() > 0) {
            outState.putParcelableArray(
                    TaggerActivity.ARG_AVAILABLE_TAGS,
                    mAvailableTags.toArray(new ITag[mAvailableTags.size()]));
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onResume() {
        super.onResume();
//        mPresenter.start();
    }


    @Override
    protected void onDestroy() {
        mPresenter.destroyView();
        mViewLoading = null;
        mSelectedTags = null;
        mAvailableTags = null;
        mAvailableTagsAdapter = null;
        mNoSelectedTags = null;
        mNoAvailableTags = null;
        mSelectedTagsAdapter = null;
        super.onDestroy();
    }


    //region MainContract.View
    /**
     * Set the presenter for the view.
     * @param presenter The presenter for this view.
     */
    @Override
    public void setPresenter(MainContract.Presenter presenter) {
        mPresenter = presenter;
    }

    /**
     * Request the loading indicator be displayed or hidden.
     * @param active If true, show the loading indicator. If false, hide it.
     */
    @Override
    public void setLoadingIndicator(boolean active) {
        mViewLoading.setVisibility(active ? View.VISIBLE : View.GONE);
    }

    /**
     * Update the view with the available tags provided.
     * @param tags The tags fetched from the Repo.
     */
    @Override
    public void setAvailableTags(@Nullable List<? extends ITag> tags) {
        Log.d(TAG, "Updating view with available tags ["
                + (tags == null ? "null" : tags.size()) + "]");
        if (tags != null) {
            mAvailableTagsAdapter.setTags(tags);
        }
        if (tags != null && tags.size() > 0) {
            showNoAvailableTags(false);
        } else {
            showNoAvailableTags(true);
        }
        mAvailableTags = tags;
    }

    /**
     * Update the view with the selected tags provided. These are the tags
     * the user selected using the custom Tag selector.
     *
     * @param tags A list of tags the user selected.
     */
    @Override
    public void setSelectedTags(@Nullable List<? extends ITag> tags) {
        Log.d(TAG, "Updating view with selected tags ["
                + (tags == null ? "null" : tags.size()) + "]");

        if (tags != null && !tags.isEmpty()) {
            showNoSelectedTags(false);
            mSelectedTagsAdapter.setTags(tags);
        } else {
            showNoSelectedTags(true);
        }
    }

    /**
     * Show the "no available tags for display" view.
     *
     * @param on True if the No Available Tags view should be displayed, else false.
     */
    @Override
    public void showNoAvailableTags(boolean on) {
        if (on) {
            mNoAvailableTags.setVisibility(View.VISIBLE);
        } else {
            mNoAvailableTags.setVisibility(View.GONE);
        }
    }

    /**
     * Show the "no selected tags for display" view.
     *
     * @param on True if the No Selected Tags view should be displayed, else false.
     */
    @Override
    public void showNoSelectedTags(boolean on) {
        if (on) {
            mNoSelectedTags.setVisibility(View.VISIBLE);
        } else {
            mNoSelectedTags.setVisibility(View.GONE);
        }
    }

    /**
     * Display an error message to the end user.
     *
     * @param msg The message to be displayed.
     */
    @Override
    public void showError(@NonNull String msg) {
        AlertUtils.displayErrorMessage(this, msg);
    }

    /**
     * @return True if the view is still active, else false.
     */
    @Override
    public boolean isActive() {
        return !AndroidUtils.hasMinimumApi(17) || !isDestroyed();
    }
    //endregion


    //region ClickListener
    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_api:
                showNoAvailableTags(true);
                mPresenter.loadTagsFromRemote();
                break;
            case R.id.btn_database:
                showNoAvailableTags(true);
                mPresenter.loadTagsFromLocal();
                break;
            case R.id.btn_selectTags:
                if (mAvailableTags != null && mAvailableTags.size() > 0) {
                    mPresenter.selectTags(this, mAvailableTags, mSelectedTags);
                } else {
                    Toast.makeText(
                            this,
                            R.string.no_tags_loaded,
                            Toast.LENGTH_SHORT).show();
                }
                break;
            default:
                Log.e(TAG, "Unknown view sent to onClick handler!");
                break;
        }
    }
    //endregion


    //region Testing
    @VisibleForTesting
    public IdlingResource getCountingIdlingResource() {
        return EspressoIdlingResource.getIdlingResource();
    }
    //endregion


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == TaggerActivity.REQUEST_CODE) {
            if (resultCode == RESULT_OK) {
                Parcelable[] pSelTags
                        = data.getParcelableArrayExtra(TaggerActivity.ARG_SELECTED_TAGS);
                mSelectedTags.clear();
                if (pSelTags != null && pSelTags.length > 0) {
                    for (Parcelable p : pSelTags) {
                        mSelectedTags.add((ITag)p);
                    }
                }
                setSelectedTags(mSelectedTags);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }
}

File: app/src/main/java/com/amandariu/tagger/demo/main/MainActivity.kt
package com.amandariu.tagger.demo.main

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.annotation.VisibleForTesting
import android.support.test.espresso.IdlingResource
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast

import com.amandariu.tagger.ITag
import com.amandariu.tagger.TaggerActivity
import com.amandariu.tagger.demo.R
import com.amandariu.tagger.demo.Injection
import com.amandariu.tagger.demo.utils.EspressoIdlingResource
import com.amandariu.tagger.demo.utils.displayErrorMessage
import com.amandariu.tagger.demo.utils.hasMinimumApi
import java.util.ArrayList

class MainActivity : AppCompatActivity(), MainContract.View, View.OnClickListener {

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }

    private var mPresenter: MainContract.Presenter? = null
    //
    // View components
    private var mAvailableTagsAdapter: SimpleTagListAdapter? = null
    private var mSelectedTagsAdapter: SimpleTagListAdapter? = null
    private var mViewLoading: ViewGroup? = null
    private var mNoAvailableTags: TextView? = null
    private var mNoSelectedTags: TextView? = null
    //endregion

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //
        // Create the presenter
        MainPresenter(Injection.provideTagsRepository(applicationContext), this)
        //
        // Initialize View Components
        //
        // Get Tags from the best available option
        val btnGetTags = findViewById<Button>(R.id.btn_loadTags)
        btnGetTags.setOnClickListener(this)
        //
        // Get Tags from API
        val btnGetApi = findViewById<Button>(R.id.btn_api)
        btnGetApi.setOnClickListener(this)
        //
        // Get Tags from Database
        val btnGetDb = findViewById<Button>(R.id.btn_database)
        btnGetDb.setOnClickListener(this)
        //
        // Open the Tag Selector
        val btnSelectTags = findViewById<Button>(R.id.btn_selectTags)
        btnSelectTags.setOnClickListener(this)
        //
        // Display Available Tags
        mNoAvailableTags = findViewById(R.id.txt_noAvailableTags)
        val availList = findViewById<RecyclerView>(R.id.list_availableTags)
        mAvailableTagsAdapter = SimpleTagListAdapter(ArrayList())
        availList.adapter = mAvailableTagsAdapter
        val clearAvailableTags = findViewById<Button>(R.id.btn_clearAvailableTags)
        clearAvailableTags.setOnClickListener(this)
        //
        // Display Selected Tags
        mNoSelectedTags = findViewById(R.id.txt_noSelectedTags)
        val selectList = findViewById<RecyclerView>(R.id.list_selectedTags)
        mSelectedTagsAdapter = SimpleTagListAdapter(ArrayList())
        selectList.adapter = mSelectedTagsAdapter
        val clearSelectedTags = findViewById<Button>(R.id.btn_clearSelectedTags)
        clearSelectedTags.setOnClickListener(this)
        //
        // Loading indicator
        mViewLoading = findViewById(R.id.view_loading)

        if (savedInstanceState != null) {
            //
            // Restore any selected tags saved during configuration change.
            val selectedTags = ArrayList<ITag>()
            val pSelTags = savedInstanceState
                    .getParcelableArray(TaggerActivity.ARG_SELECTED_TAGS)
            if (pSelTags != null && pSelTags.size > 0) {
                for (t in pSelTags) {
                    selectedTags.add(t as ITag)
                }
                setSelectedTags(selectedTags)
            }
            //
            // Restore available tags saved during configuration change
            val pAvaTags = savedInstanceState
                    .getParcelableArray(TaggerActivity.ARG_AVAILABLE_TAGS)
            if (pAvaTags != null && pAvaTags.size > 0) {
                val tags = ArrayList<ITag>(pAvaTags.size)
                for (t in pAvaTags) {
                    tags.add(t as ITag)
                }
                setAvailableTags(tags)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        //
        // Save selected tags
        val selectedTags = mSelectedTagsAdapter!!.tags
        if (selectedTags.size > 0) {
            outState!!.putParcelableArray(
                    TaggerActivity.ARG_SELECTED_TAGS,
                    selectedTags.toTypedArray())
        }
        //
        // Save available tags
        val availableTags = mAvailableTagsAdapter!!.tags
        if (availableTags.size > 0) {
            outState!!.putParcelableArray(
                    TaggerActivity.ARG_AVAILABLE_TAGS,
                    availableTags.toTypedArray())
        }
        super.onSaveInstanceState(outState)
    }

    override fun onResume() {
        super.onResume()
        //        mPresenter.start();
    }

    override fun onDestroy() {
        mPresenter!!.destroyView()
        mViewLoading = null
        mAvailableTagsAdapter = null
        mNoSelectedTags = null
        mNoAvailableTags = null
        mSelectedTagsAdapter = null
        super.onDestroy()
    }

    //region MainContract.View
    /**
     * Set the presenter for the view.
     * @param presenter The presenter for this view.
     */
    override fun setPresenter(presenter: MainContract.Presenter) {
        mPresenter = presenter
    }

    /**
     * Request the loading indicator be displayed or hidden.
     * @param active If true, show the loading indicator. If false, hide it.
     */
    override fun setLoadingIndicator(active: Boolean) {
        mViewLoading!!.visibility = if (active) View.VISIBLE else View.GONE
    }

    /**
     * Update the view with the available tags provided.
     * @param tags The tags fetched from the Repo.
     */
    override fun setAvailableTags(tags: List<ITag>?) {
        Log.d(TAG, "Updating view with available tags ["
                + (tags?.size ?: "null") + "]")
        if (tags != null) {
            mAvailableTagsAdapter!!.setTags(tags)
        }
        if (tags != null && tags.size > 0) {
            showNoAvailableTags(false)
        } else {
            showNoAvailableTags(true)
        }
    }

    /**
     * Update the view with the selected tags provided. These are the tags
     * the user selected using the custom Tag selector.
     *
     * @param tags A list of tags the user selected.
     */
    override fun setSelectedTags(tags: List<ITag>?) {
        Log.d(TAG, "Updating view with selected tags ["
                + (tags?.size ?: "null") + "]")

        if (tags != null && !tags.isEmpty()) {
            showNoSelectedTags(false)
            mSelectedTagsAdapter!!.setTags(tags)
        } else {
            showNoSelectedTags(true)
        }
    }

    /**
     * Show the "no available tags for display" view.
     *
     * @param on True if the No Available Tags view should be displayed, else false.
     */
    override fun showNoAvailableTags(on: Boolean) {
        if (on) {
            mAvailableTagsAdapter!!.clearTags()
            mNoAvailableTags!!.visibility = View.VISIBLE
        } else {
            mNoAvailableTags!!.visibility = View.GONE
        }
    }

    /**
     * Show the "no selected tags for display" view.
     *
     * @param on True if the No Selected Tags view should be displayed, else false.
     */
    override fun showNoSelectedTags(on: Boolean) {
        if (on) {
            mNoSelectedTags!!.visibility = View.VISIBLE
        } else {
            mNoSelectedTags!!.visibility = View.GONE
        }
    }

    /**
     * Display an error message to the end user.
     *
     * @param msg The message to be displayed.
     */
    override fun showError(msg: String) {
        displayErrorMessage(this, msg)
    }

    /**
     * @return True if the view is still active, else false.
     */
    override val isActive: Boolean
        get() = !hasMinimumApi(17) || !isDestroyed

    //endregion

    //region ClickListener
    override fun onClick(view: View) {
        when (view.id) {
            R.id.btn_loadTags -> {
                showNoAvailableTags(true)
                mPresenter!!.loadTags(false)
            }
            R.id.btn_api -> {
                showNoAvailableTags(true)
                mPresenter!!.loadTagsFromRemote(this@MainActivity)
            }
            R.id.btn_database -> {
                showNoAvailableTags(true)
                mPresenter!!.loadTagsFromLocal()
            }
            R.id.btn_selectTags -> {
                val availableTags = mAvailableTagsAdapter!!.tags
                val selectedTags = mSelectedTagsAdapter!!.tags
                if (availableTags.size > 0) {
                    mPresenter!!.selectTags(this, availableTags, selectedTags)
                } else {
                    Toast.makeText(
                            this,
                            R.string.no_tags_loaded,
                            Toast.LENGTH_LONG).show()
                }
            }
            R.id.btn_clearAvailableTags -> {
                mAvailableTagsAdapter!!.clearTags()
                showNoAvailableTags(true)
            }
            R.id.btn_clearSelectedTags -> {
                mSelectedTagsAdapter!!.clearTags()
                showNoSelectedTags(true)
            }
            else -> Log.e(TAG, "Unknown view sent to onClick handler!")
        }
    }
    //endregion

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode == TaggerActivity.REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                val pSelTags = data.getParcelableArrayExtra(TaggerActivity.ARG_SELECTED_TAGS)
                if (pSelTags != null && pSelTags.size > 0) {
                    val selectedTags = ArrayList<ITag>()
                    for (p in pSelTags) {
                        selectedTags.add(p as ITag)
                    }
                    setSelectedTags(selectedTags)
                }
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    //region Testing
    @Suppress("UNUSED")
    val countingIdlingResource: IdlingResource
        @VisibleForTesting
        get() = EspressoIdlingResource.idlingResource
    //endregion
}
